### PR TITLE
[SD-XL] Add inpainting

### PR DESCRIPTION
**Pull request by patrickvonplaten**
## SD-XL inpainting

This PR solves: https://github.com/huggingface/diffusers/issues/4080 and is ready for a review.

Inpainting works well for both the vanilla case and the "Ensemble of Expert Denoisers case".

You can try the following to see for yourself:

### Vanilla inpainting:

```py
import torch
from diffusers import StableDiffusionXLInpaintPipeline
from diffusers.utils import load_image

pipe = StableDiffusionXLInpaintPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-0.9", torch_dtype=torch.float16, variant="fp16", use_safetensors=True
)
pipe.to("cuda")

img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

init_image = load_image(img_url).convert("RGB")
mask_image = load_image(mask_url).convert("RGB")

prompt = "A red cat sitting on a bench"
image = pipe(prompt=prompt, image=init_image, mask_image=mask_image, num_inference_steps=50, strength=0.80).images[0]
```

### Ensemble of Expert of denoisers 

which should give slightly better quality:

```py
from diffusers import StableDiffusionXLInpaintPipeline
from diffusers.utils import load_image

pipe = StableDiffusionXLInpaintPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-0.9", torch_dtype=torch.float16, variant="fp16", use_safetensors=True
)
pipe.to("cuda")

refiner = StableDiffusionXLInpaintPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-refiner-0.9",
    text_encoder_2=pipe.text_encoder_2,
    vae=pipe.vae,
    torch_dtype=torch.float16,
    use_safetensors=True,
    variant="fp16",
)
refiner.to("cuda")

img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

init_image = load_image(img_url).convert("RGB")
mask_image = load_image(mask_url).convert("RGB")

prompt = "A red cat sitting on a bench"
num_inference_steps = 75
high_noise_frac = 0.7

image = pipe(
    prompt=prompt,
    image=init_image,
    mask_image=mask_image,
    num_inference_steps=num_inference_steps,
    strength=0.80,
    denoising_start=high_noise_frac,
    output_type="latent",
).images
image = refiner(
    prompt=prompt,
    image=image,
    mask_image=mask_image,
    num_inference_steps=num_inference_steps,
    denoising_start=high_noise_frac,
).images[0]
```
aaa

### CODE REVIEW

**Comment by gkorepanov** [comment]
Hi, is there already a SD XL checkpoint with unet having 9 channels? Seems like no specific inpainting model was released for SD XL, but without it inpainting results are meaningless (I mean there is little to no semantic match between inpainted regions and present regions in generated images)

**Comment by AmericanPresidentJimmyCarter** [comment]
Is this the same method as `InpaintLegacy` in SD? There are now 3 inpainting methods for LDM, the "InpaintLegacy" method, the model using extra channels from RunwayML, and PSLD. I think we should maintain consistent naming.

**Comment by AmericanPresidentJimmyCarter** [comment]
And yes, I agree with @gkorepanov , the "InpaintLegacy" method is more or less useless.

**Comment by patrickvonplaten** [comment]
`StableDiffusionInpaintPipelineLegacy` is deprecated and will be removed. Everything you were able to do with `StableDiffusionInpaintPipelineLegacy` you can now do with `StableDiffusionInpaintPipeline`

In that sense there will  only be one "true" `StableDiffusionXLInpaintPipeline`

**Comment by adhikjoshi** [comment]
As inpaiting checkpoint isn't there, does it affects quality in general?

**Comment by patrickvonplaten** [comment]
> As inpaiting checkpoint isn't there, does it affects quality in general?

Works pretty well for me for now, I recommend making sure to pass `strength=0.7` or `strength=0.8`.

I think the checkpoint will however have problems when you want to replace the masked area with something very different to what was there before.

File: docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.mdx [hunk 1]
```diff
@@ -57,6 +57,50 @@ prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"
...
+
+You can use SDXL as follows for *image-to-image*:
+
+```py 
+import torch
+from diffusers import StableDiffusionXLImg2ImgPipeline
+from diffusers.utils import load_image
+
+pipe = StableDiffusionXLImg2ImgPipeline.from_pretrained(
+    "stabilityai/stable-diffusion-xl-refiner-0.9", torch_dtype=torch.float16, variant="fp16", use_safetensors=True
```
Comment (nhnt11) [inline docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.mdx:70]: Is it really intended to use the refiner model for general img2img? I've been trying to understand this - I've also seen it [here](https://huggingface.co/docs/diffusers/api/pipelines/stable_diffusion/stable_diffusion_xl#imagetoimage) for example - but I think I am missing something. My understanding is that the refiner model is intended as a kind of de-noising and/or fidelity-increasing step and it isn't good at generating the kind of baseline content of the image. If that's correct, feels like it'd perform poorly for img2img with lower strength values.

Comment (patrickvonplaten) [reply docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.mdx:70]: You can use both! The refiner might be better suited for images that look already like the prompt which is the case here. We should maybe improve the docs after the official release.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 2]
```diff
@@ -981,8 +981,6 @@ def __call__(
             generator,
             do_classifier_free_guidance,
         )
-        init_image = init_image.to(device=device, dtype=masked_image_latents.dtype)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:984]: This is never used actually and was a copy-paste bug I think

File: docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.mdx (hunk 1 above)
Comment (nhnt11) [reply docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.mdx:70]: Gotcha - thanks for clarifying!
